docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -0,0 +1,312 @@
+
+ADR: This is good info, should probably go at the beginning,
+"These are our decisions on what to implement and what not to.
+The rest of the document only tackles the items reported as "to implement" here.
+
+### Lane coordination
+
+There should be some kind of Lane Coordinator that receives uncategorised txs and assigns them a lane.
+This could be implemented in the reactor or inside `CListMempool`.
+* For an MVP, probably it's easier to put the coordination logic in `CListMempool`.
+* There will be a _native_ or _default_ lane, similar to the _native VLAN_ in an 802.1Q network.
+* `CheckTx` will (optionally) return lane information, and only then it can be put in a lane.
+  (So a tx in a lane has been validated at least once by some node.)
+  * If `CheckTx` provides no lane information: native lane
+* No difference in treatment between TXs coming from RPC, or TXs coming from peers (p2p)
+  * Peers aren't trusted
+  * In any case, the TXs need to be validated first via `CheckTx`
+* The duality lane/priority introduces a powerful indirection
+  * The app can just define the lane of a transaction in `CheckTx`, but the priority of the lane
+    itself can configured (and fine-tuned) elsewhere (also the app?, operators?)
+* Two options for introducing several TX lists:
+  1. Multiple `Clist`s inside `CListMempool`, the cache stays inside `CListMempool`.
+  2. Multiple `CListMempool` instances, the cache needs to be outside `CListMempool` (likely, directly controlled by the reactor)
+* We need to investigate the details of these two options in terms of complexity and risks in the code
+
+ADR: Probably part of [Data flow](#data-flow) section:
+1. some bullets are clearly "Entry flow"
+2. other are general considerations of all flows, or architectural.
+   These are likely to go in a first, introductory subsection of [Data flow](#data-flow) section
+
+### Data flow
+
+* Entry flow:
+  * [Rough description]: the steps when a Tx enters the node are unmodified, except
+    * `CheckTx` (optionally) returns lane information (if TX valid)
+    * The lane information is used to decide to which transaction list the Tx will be added
+* Broadcast flow:
+  * Higher-priority lanes get to be gossiped first, and gossip within a lane is still in FIFO order.
+  * [Rough description]: we need to extend the loop in the broadcast goroutine. Several channels now, not just one
+    * To decide
+      * if we're going for (simpler) solution where with max 2 lanes,
+        the broadcast goroutine itself could manage the priority of lanes within the existing `select` at the end of the loop
+      * if we're going for (future-proof) solution where max lanes can be configurable,
+        we need to refactor method `TxWaitChan` (or the code that sends the data to that channel) to manage the lane priority there.
+        In this case, the broadcast goroutine will be mainly unmodified.
+* Reap flow
+  * TXs are reaped from higher-priority lanes first, in FIFO order.
```

Comment:
There is a logic for this in `func (c *MConnection) sendPacketMsg() bool` in `p2p/conn` package. The highest priority first approach may lead to starvation in low priority lanes.

Reply:
> The highest priority first approach may lead to starvation in low priority lanes

Yes, just like QoS traffic classes in IP networks. When speaking about QoS, it comes with the territory.

Reply:
About the logic in p2p implementing priority in channels, thanks for pointing out; @hvanz we'll need to take a look to see how we can leverage that in our design.